### ISSUE TITLE
rustic-cargo-run throws error could not choose rustc version.

### ISSUE THREAD

**Issue by apoorv569** [brotzeit/rustic#444]
I recently started learning `rust` and installed `rustic` mode. I can do `cargo init` `cargo run` and `rustup` commands fine from terminal, but when I try to run most `rustic` or `cargo` related commands in `Emacs` I get error similar as the title of this issue. I am pretty sure I need to change some variables as I have changed my `$CARGO_HOME` and `$RUSTUP_HOME` to `~/.loca/share/cargo` and `~/.local/share/rustup` respectively, but I am not sure which variables as I have tried modifying several but failed to fix the error. If someone can kindly help me fix this issue it would be nice.

Here is my `rustic` configuration,
```elisp
  (use-package rustic
    :straight t
    :hook (rustic-mode . lsp-deferred)
    :config
    (with-eval-after-load "lsp-mode"
      (add-to-list 'lsp-enabled-clients 'rust-analyzer))
    :custom
    (rustic-cargo-bin "~/.local/share/rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo"))
```

I have these `rustup` components already installed,
```shell
❯ rustup component list | grep installed
cargo-x86_64-unknown-linux-gnu (installed)
clippy-x86_64-unknown-linux-gnu (installed)
rls-x86_64-unknown-linux-gnu (installed)
rust-analysis-x86_64-unknown-linux-gnu (installed)
rust-docs-x86_64-unknown-linux-gnu (installed)
rust-src (installed)
rust-std-x86_64-unknown-linux-gnu (installed)
rustc-x86_64-unknown-linux-gnu (installed)
rustc-dev-x86_64-unknown-linux-gnu (installed)
rustfmt-x86_64-unknown-linux-gnu (installed)
```

and I have the following toolchain installed and set as default,
```shell
❯ rustup toolchain list
stable-x86_64-unknown-linux-gnu (default)

~
❯ rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/apoorv/.local/share/rustup

stable-x86_64-unknown-linux-gnu (default)
rustc 1.63.0 (4b91a6ea7 2022-08-08)
```

**Comment by psibi** [comment]
What error does it throw, can you paste the log ?

Also, what happens if you change your configuration to this:

```
  (use-package rustic
    :straight t
    :hook (rustic-mode . lsp-deferred)
    :config
    (with-eval-after-load "lsp-mode"
      (add-to-list 'lsp-enabled-clients 'rust-analyzer)))
```

**Comment by psibi** [comment]
@apoorv569 Can you show the ouptput of `rustup show` ?

Also, can you do the following steps:

1. Do `rustup default stable`
2. Then repeat the same commands with rustic and see if the error is still reproducible.

**Comment by apoorv569** [comment]
> @apoorv569 Can you show the ouptput of `rustup show` ?

I did share the output of `rustup show` in the previous reply.

> 
> Also, can you do the following steps:
> 
>     1. Do `rustup default stable`
> 
>     2. Then repeat the same commands with rustic and see if the error is still reproducible.

I have already tried running `rustup default stable`, but no luck, I still get the same error.

**Comment by psibi** [comment]
Can you run `cargo run` outside of Emacs ? If yes, does that work fine ?

**Comment by apoorv569** [comment]
> Can you run `cargo run` outside of Emacs ? If yes, does that work fine ?

Yes, I already mentioned this in the original post as well.

**Comment by psibi** [comment]
Ah, I see. Unfortunately I think this is something to do with the setup. I'm unable to reproduce it locally and I'm not sure if I can help further unless I have a way to reproduce this.

**Comment by apoorv569** [comment]
> Ah, I see. Unfortunately I think this is something to do with the setup. I'm unable to reproduce it locally and I'm not sure if I can help further unless I have a way to reproduce this.

Yes, probably because you didn't change the default `PATH` for `CARGO_HOME` and `RUSTUP_HOME`.

**Comment by apoorv569** [comment]
I managed to fix the issue, the problem was my personal `emacs` configuration was updating `exec-path` variable from my `shell` `PATH` variable, so it couldn't find the binaries required for this setup to work.

I installed [this](https://github.com/purcell/exec-path-from-shell) package and it fixed the problem for me.

Here is the configuration part I added.
```elisp
  (use-package exec-path-from-shell
    :straight t)

  (when (memq window-system '(mac ns x))
    (exec-path-from-shell-initialize))
```

I am closing this issue now. Thanks!
